feat: cssc patch command to enable continuous patching

Ruchii-27 opened the pull request:
**Purpose of the PR**

- Added new cssc patch commands which would be instrumental in enabling continuous patching on the registry. These commands perform filtration either using filter policy or from a filter file in Json format from local file system if dry run is chosen.
- In this PR, only dry run functionality is enabled. The actual patch implementation will be added in upcoming PRs.
- Refactored current list tag implementation to return a tag list so that it can be re-used both in tag and cssc commands + updated corresponding tests
- Added a bunch of unit tests covering both success and failure scenarios for the cssc command
- Refactored and restructured to move non cmd related methods to internal package for cssc and tag
- Verified by running changes locally ensuring all unit tests pass and also via ACR task

sajayantony commented on cmd/acr/cssc.go:
```diff
@@ -0,0 +1,151 @@
+
+const (
+	newCsscCmdLongMessage  = `acr cssc: CSSC (Container Secure Supply Chain) operations for a registry. Use subcommands for more specific operations.`
+	newPatchCmdLongMessage = `acr cssc patch: List all repositories and tags that match the filter.
```
Add the preview label to these commands so that its not indicated as GA. 
For e.g. https://github.com/oras-project/oras/blob/main/cmd/oras/root/discover.go#L49

Ruchii-27 replied:
Addressed this in the latest commit.

estebanreyl commented on internal/cssc/cssc.go:
```diff
@@ -0,0 +1,190 @@
+
+// Reads the filter policy from the specified repository and tag and returns the Filter struct
+func GetFilterFromFilterPolicy(ctx context.Context, filterPolicy string, loginURL string, username string, password string) (Filter, error) {
+	if !strings.Contains(filterPolicy, ":") {
```
I would suggest doing a regex check to validate the input upfront. Repositories have certain naming limitations as do tags. This check will allow things with more than one : which is invalid for example.

estebanreyl commented on internal/cssc/cssc.go:
```diff
@@ -0,0 +1,190 @@
+	// Connect to the remote repository
+	repo, err := remote.NewRepository(fmt.Sprintf("%s/%s", loginURL, filterRepoName))
+	if err != nil {
+		panic(err)
```
You should not be panicking here since the command itself can propagate and possibly return the error to the user. Instead wrap the error and return it up.

Ruchii-27 replied:
Addressed this in the latest commit.

estebanreyl commented on internal/cssc/cssc.go:
```diff
@@ -0,0 +1,190 @@
+	if err != nil {
+		return Filter{}, errors.Wrap(err, "error fetching manifest content when reading the filter policy")
+	}
+	var pulledManifest v1.Manifest
```
Manifests can also be manifest lists which do not have layers. Do you also need to account for that scenario or is the filter policy guaranteed to be a regular oci manifest?

Ruchii-27 replied:
The json filter file is always uploaded as an OCI artifact using Oras.

estebanreyl commented on internal/cssc/cssc.go:
```diff
@@ -0,0 +1,190 @@
+	for _, layer := range pulledManifest.Layers {
+		fileContent, err = content.FetchAll(ctx, repo, layer)
+		if err != nil {
+			panic(err)
```
Same thing, propagate the error

Ruchii-27 replied:
sure, done in the latest commit.

estebanreyl commented on internal/cssc/cssc.go:
```diff
@@ -0,0 +1,190 @@
+		if filterRepo.Enabled != nil && !*filterRepo.Enabled {
+			continue
+		}
+		if filterRepo.Repository == "" || filterRepo.Tags == nil || len(filterRepo.Tags) == 0 || filterRepo.Tags[0] == "" {
```
filterRepo.Tags[0] == "" is this something that can happen in practice? Why might the first tag entry be empty, should that validation happen at a tag level?

Ruchii-27 replied:
No, it won't happen in practice. Repository and tags are required attributes in json filter and this validation will happen when the customer attempts to upload it using az cli cssc commands. Ideally none of these conditions should be true in practice. I'll remove this last condition filterRepo.Tags[0] == "", however will continue to keep others for gracefully skipping and continue.

estebanreyl commented on internal/cssc/cssc.go:
```diff
@@ -0,0 +1,190 @@
+
+		tagList, err := tag.ListTags(ctx, acrClient, filterRepo.Repository)
+		if err != nil {
+			if strings.Contains(err.Error(), "404") {
```
As a general consideration, other errors may be caught in this type of check (e.g error in line 404), it may not be a big deal if you own the other function but its still best practice to avoid this . In golang we try to check for the error type taking advantage of the wrap functionality [as seen here](https://gosamples.dev/check-error-type/). Generally Id use an errors.As(err, &badInputErr) to check for a specific error type, if possible,

Ruchii-27 replied:
ListTags is a method we own, however it was not returning a specific type of error.
Updated it to do that and updated the code to use errors.As.

estebanreyl commented on internal/cssc/cssc.go:
```diff
@@ -0,0 +1,190 @@
+
+		if len(filterRepo.Tags) == 1 && filterRepo.Tags[0] == "*" {
+			for _, tag := range tagList {
+				if strings.Contains(*tag.Name, "-patched") {
```
I am guessing patched is a suffix, right? Can we instead use a has suffix which should be more efficient or is there more after -patched? Also is this to be a reserved tag name?

Ruchii-27 replied:
updated to use HasSuffix. Yes, in the 1st phase, patched image will always have a tag ending with -patched.

estebanreyl commented on internal/cssc/cssc.go:
```diff
@@ -0,0 +1,190 @@
+				}
+			}
+		} else {
+			for _, ftag := range filterRepo.Tags {
```
The logic here is not that clear to me from the code.  I am not sure clear on what the ftag is expected to be like. Do we have a spec on this for my reference?

Ruchii-27 replied:
I'll add some comments in there to make it clearer. Yes, we do have a spec, I'll share it shortly.

estebanreyl commented on internal/cssc/cssc.go:
```diff
@@ -0,0 +1,190 @@
+		fmt.Println("No matching repository and tag found!")
+	}
+	if showPatchTags {
+		fmt.Println("Listing repositories and tags matching the filter with corrosponding patch tag (if present):")
```
Nit: corresponding

Ruchii-27 replied:
updated

estebanreyl commented on internal/cssc/cssc.go:
```diff
@@ -0,0 +1,190 @@
+	if showPatchTags {
+		fmt.Println("Listing repositories and tags matching the filter with corrosponding patch tag (if present):")
+		for _, result := range filteredResult {
+			fmt.Printf("%s/%s:%s,%s\n", loginURL, result.Repository, result.Tag, result.PatchTag)
```
Is there a need to include the loginurl for each of these?

Ruchii-27 replied:
yeah, I've done that to keep the output consistent with other pre-existing commands like acr tag list

estebanreyl commented on internal/cssc/cssc.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+// Appends the element to the slice if it does not already exist in the slice
+func AppendElement(slice []FilteredRepository, element FilteredRepository) []FilteredRepository {
```
This is not efficient. Use a set or I suppose in go a map[string]struct{} since go technically doesn't have sets.

Ruchii-27 replied:
I realized I had a double loop there. Simplified it now.

Ruchii-27 replied:
Updated the code to update the slice element if it exists. Thanks for the suggestion and help.

estebanreyl commented on internal/cssc/cssc_test.go:
```diff
@@ -0,0 +1,280 @@
+		assert.ErrorContains(t, err, "failed getting the tags")
+		assert.Nil(t, filteredRepositories)
+	})
+	//4. If filter has a tag that doesn't exist in the repository, ignore it and return whatever exists that matches the filter
```
Nit. Use consistent spacing in go comments, should be // 4

Ruchii-27 replied:
updated

estebanreyl commented on internal/tag/tag.go:
```diff
@@ -0,0 +1,58 @@
+)
+
+// ListTags will do the http requests and return the digest of all the tags in the selected repository.
+func ListTags(ctx context.Context, acrClient api.AcrCLIClientInterface, repoName string) ([]acr.TagAttributesBase, error) {
```
Nit. ListAllTags

Ruchii-27 replied:
Changing the method name required me to do renaming in purge cmds and related test files. I'll skip this change to limit the scope of changes to cssc for now.

estebanreyl commented on internal/tag/tag.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+// DeleteTags receives an array of tags digest and deletes them using the supplied acrClient.
+func DeleteTags(ctx context.Context, acrClient api.AcrCLIClientInterface, loginURL string, repoName string, args []string) error {
```
Nit. What is args, it looks like it is a tag list? Let's make sure the param name is self explanatory

Ruchii-27 replied:
Addressed this.

estebanreyl reviewed (commented):
Overall I am a little confused about some of the finer points of what the patch files look like and what the plan is. Also left some comments on some optimizations we can do for the code.

estebanreyl commented on internal/cssc/cssc.go:
```diff
@@ -0,0 +1,190 @@
+		if len(filterRepo.Tags) == 1 && filterRepo.Tags[0] == "*" {
+			for _, tag := range tagList {
+				if strings.Contains(*tag.Name, "-patched") {
+					originalTag := strings.Split(*tag.Name, "-patched")[0]
```
Nit. Since -patched is a suffix a more efficient approach would be to get a substring from [0 : len(str)- len("-patched")-1] to avoid the string search

Ruchii-27 replied:
Updated as per your suggestion

Ruchii-27 commented on cmd/acr/cssc.go:
```diff
@@ -0,0 +1,151 @@
+
+const (
+	newCsscCmdLongMessage  = `acr cssc: CSSC (Container Secure Supply Chain) operations for a registry. Use subcommands for more specific operations.`
+	newPatchCmdLongMessage = `acr cssc patch: List all repositories and tags that match the filter.
```
Updated the description. This command is meant to patch, however actual patch implementation will come in later. Right now, the goal is to only support dry-run which can list repositories and tags matching the filter.

estebanreyl replied:
Thats fine but you should mark that as preview then and note there will be a functionality change.

estebanreyl replied:
Looks good

Ruchii-27 commented on cmd/acr/cssc.go:
```diff
@@ -0,0 +1,151 @@
+func newPatchFilterCmd(csscParams *csscParameters) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "patch",
+		Short: "Manage cssc patch operations for a registry",
```
Addressed this in the latest commit.

Ruchii-27 commented on cmd/acr/cssc.go:
```diff
@@ -0,0 +1,151 @@
+				return err
+			}
+			loginURL := api.LoginURL(registryName)
+			getRegistryCredsFromStore(csscParams, loginURL)
```
Updated the method name in latest commit

Ruchii-27 commented on cmd/acr/cssc.go:
```diff
@@ -0,0 +1,151 @@
+				return errors.New("flag --file-path can only be used in combination with --dry-run")
+			} else if csscParams.dryRun {
+				if csscParams.filePath == "" {
+					return errors.New("flag --file-path is required when using --dry-run")
```
I've updated the design to simplify things. dry-run will now be supported both for filter-policy as well as for filter-file-path.

Ruchii-27 commented on cmd/acr/cssc.go:
```diff
@@ -0,0 +1,151 @@
+
+			if len(filter.Repositories) == 0 {
+				fmt.Println("Filter is empty or invalid.")
+				return nil
```
I verified this, error messages are shown to the user.

Ruchii-27 commented:
> Overall, I am a little confused about some of the finer points of what the patch files look like and what the plan is. Also left some comments on some optimizations we can do for the code.

Thanks for the review Esteban! I've addressed all of your comments and also simplified the design to make things more clear and less confusing in my latest commit.

estebanreyl commented on cmd/acr/cssc.go:
```diff
@@ -0,0 +1,151 @@
+		},
+	}
+
+	cmd.PersistentFlags().StringVar(&csscParams.filterPolicy, "filter-policy", "", "The filter policy defined by the filter json file uploaded in a repo:tag. For v1, it should be continuouspatchpolicy:latest")
```
sure but didn't you note that the filter-policy should be continuouspatchpolicy:v1?

estebanreyl replied:
discussed offline, keep as is

estebanreyl commented:
test

estebanreyl reviewed (approved):
LGTM
